### PR TITLE
fix(authz_mnesia): add a soft limit in the API for the length of ACL rules

### PR DESCRIPTION
**Pull request by lafirest**
Fixes EMQX-12465
Release version: v/e5.?

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx_auth_mnesia/test/emqx_authz_api_mnesia_SUITE.erl [hunk 1]
```diff
@@ -36,7 +36,7 @@ init_per_suite(Config) ->
             {emqx_conf,
                 "authorization.cache { enable = false },"
                 "authorization.no_match = deny,"
-                "authorization.sources = [{type = built_in_database}]"},
+                "authorization.sources = [{type = built_in_database, max_rules_len = 5}]"},
```
Comment (zmstone) [inline apps/emqx_auth_mnesia/test/emqx_authz_api_mnesia_SUITE.erl:39]:
```suggestion
                "authorization.sources = [{type = built_in_database, max_rules = 5}]"},
```

File: apps/emqx_auth_mnesia/src/emqx_authz_api_mnesia.erl [hunk 2]
```diff
@@ -700,28 +721,45 @@ rules_example({ExampleName, ExampleType}) ->
...
-binjoin([Bin]) ->
-    Bin;
-binjoin(Bins) ->
-    binjoin(Bins, <<>>).
+ensure_rules_len(Rules, MaxLen) ->
+    case erlang:length(Rules) =< MaxLen of
+        true ->
+            ok;
+        _ ->
+            {error, rules_too_long}
```
Comment (zmstone) [inline apps/emqx_auth_mnesia/src/emqx_authz_api_mnesia.erl:737]:
```suggestion
            {error, too_many_rules}
```

File: apps/emqx_auth_mnesia/src/emqx_authz_mnesia_schema.erl [hunk 3]
```diff
@@ -30,12 +30,24 @@
...
+
 namespace() -> "authz".
 
 type() -> ?AUTHZ_TYPE.
 
 fields(builtin_db) ->
-    emqx_authz_schema:authz_common_fields(?AUTHZ_TYPE).
+    emqx_authz_schema:authz_common_fields(?AUTHZ_TYPE) ++
+        [
+            {max_rules_len,
```
Comment (zmstone) [inline apps/emqx_auth_mnesia/src/emqx_authz_mnesia_schema.erl:42]:
```suggestion
            {max_rules,
```

File: rel/i18n/emqx_authz_mnesia_schema.hocon [hunk 4]
```diff
@@ -6,4 +6,7 @@ builtin_db.desc:
 builtin_db.label:
 """Builtin Database"""
 
+max_rules_len.desc:
+"""Maximum rule length per client/user. Note that performance may decrease as rule length increases."""
```
Comment (zmstone) [inline rel/i18n/emqx_authz_mnesia_schema.hocon:10]:
```suggestion
"""Maximum number of rules per client/user. Note that performance may decrease as number of rules increases."""
```

File: changes/ce/fix-13196.en.md [hunk 5]
```diff
@@ -0,0 +1 @@
+In the built-in database of authorization, added a limit for the length of rules per client/user, and the default values is 100.
```
Comment (zmstone) [inline changes/ce/fix-13196.en.md:1]:
```suggestion
In the built-in database of authorization, added a limit for the number of rules per client/user, and the default values is 100.
```

**Review by HJianBo — APPROVED** [review]
LGTM

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-3369)

**Pull request merged**